### DISCUSSION TITLE
Passing an array into TypedSQL query

### DISCUSSION THREAD

**Discussion by zdychacek** [prisma/prisma#25188] [Q&A]
### Question

Hi, thank you for the TypedSQL feature I am trying to use now, but getting an error when passing and array of integers. Can I ask for pointing out what am I doing wrong?

This is my query in .sql file:

```sql
-- @param {Int} $1:jobLevelId
-- @param {Int} $2:userId
-- @param {Int} $3:strengthThreshold
-- @param {Int} $4:feedbacksCountThreshold
-- @param {Int[]} $5:higherOrEqualLevelIds

SELECT 
    u.id AS "userId",
    u."jobLevelId",
    cf."competencyId",
    AVG(cf.value) AS "averageValue",
    COUNT(cf.id) AS "totalCount"
FROM
    "CompetencyMentorshipApplication" cma
JOIN
    "User" u ON u.id = cma."userId"
JOIN
    "CompetencyFeedback" cf ON cf."competencyId" = cma."competencyId"
JOIN
    "Feedback" f ON cf."feedbackId" = f."id"
JOIN
    "_ReceivedByUsers" rbu ON f."id" = rbu."A" AND rbu."B" = u.id
LEFT JOIN
    "Level" l ON u."jobLevelId" = l.id
WHERE
    cma."userId" != $2
    AND ($1 IS NULL OR u."jobLevelId" = $1)
    AND (ARRAY_LENGTH($5, 1) IS NULL OR u."jobLevelId" = ANY($5::integer[]))
    AND (ARRAY_LENGTH($5, 1) IS NULL OR cf."levelId" = ANY($5::integer[]))
GROUP BY
    u.id, u."jobLevelId", u."lastName", cf."competencyId", l."order"
HAVING
    AVG(cf.value) >= $3
    AND COUNT(cf.id) >= $4;
```

And this an error I am getting when running `prisma generate --sql`:

```
Environment variables loaded from .env
Prisma schema loaded from prisma/schema.prisma
Error: Errors while reading sql files:

In prisma/sql/listCompetencyMentors.sql:
Error: ERROR: could not determine polymorphic type because input has type unknown
```

Thank you very much.

Ondrej

**Comment by ludralph** [comment]
Hi @zdychacek 👋 

Thank you for raising this question.

Can you please share the schema file? meanwhile can you try this
```sql
-- @param {Int} $1:jobLevelId
-- @param {Int} $2:userId
-- @param {Int} $3:strengthThreshold
-- @param {Int} $4:feedbacksCountThreshold
-- @param {Int[]} $5:higherOrEqualLevelIds

SELECT 
    u.id AS "userId",
    u."jobLevelId",
    cf."competencyId",
    AVG(cf.value) AS "averageValue",
    COUNT(cf.id) AS "totalCount"
FROM
    "CompetencyMentorshipApplication" cma
JOIN
    "User" u ON u.id = cma."userId"
JOIN
    "CompetencyFeedback" cf ON cf."competencyId" = cma."competencyId"
JOIN
    "Feedback" f ON cf."feedbackId" = f."id"
JOIN
    "_ReceivedByUsers" rbu ON f."id" = rbu."A" AND rbu."B" = u.id
LEFT JOIN
    "Level" l ON u."jobLevelId" = l.id
WHERE
    cma."userId" != $2
    AND ($1 IS NULL OR u."jobLevelId" = $1)
    AND (ARRAY_LENGTH($5::integer[], 1) IS NULL OR u."jobLevelId" = ANY($5::integer[]))
    AND (ARRAY_LENGTH($5::integer[], 1) IS NULL OR cf."levelId" = ANY($5::integer[]))
GROUP BY
    u.id, u."jobLevelId", u."lastName", cf."competencyId", l."order"
HAVING
    AVG(cf.value) >= $3
    AND COUNT(cf.id) >= $4;
```
In this modified version, the array parameter `$5` is explicitly cast to integer[] in the ARRAY_LENGTH and ANY functions. Do let us know if this resolves the issue for you.

**Reply by Timmitry** [reply]
I don't think that this was the problem here - I'm trying something similar, and when generating the prisma client, I'm getting the error:

```
Error: SQL documentation parsing: invalid type: 'String[]' (accepted types are: 'Int', 'BigInt', 'Float', 'Boolean', 'String', 'DateTime', [....] ) at '{String[]} $1:importIds'.
```

The readme also doesn't mention any possibility to specify the input type as an array.

**Reply by nikolas-beckel** [reply]
Hey, I have the same issue. I don't know how to specify array type in the SQL file. I'm also getting the error above.
@Timmitry did you find a solution?

**Reply by Timmitry** [reply]
@nikolas-beckel No, I suspect that this simply was never implemented, as neither the documentation nor the error message suggest anything like it (the docs contain instructions about passing arrays, but not about typing them)...

**Reply by nikolas-beckel** [reply]
Hey @Timmitry, I got the answer.
https://x.com/MarcHessDev/status/1949903028133515409
<img width="605" height="333" alt="image" src="https://github.com/user-attachments/assets/252836c1-0b07-49e0-9b51-42ab75e3aaa9" />

**Reply by Timmitry** [reply]
Interesting 👀 We decided against using TypedSQL for now, so I can't check whether this works at the moment, but thanks for asking and sharing!

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that haven’t had any recent activity and appear to be outdated.

If this discussion is still important to you or unresolved, we’d love to hear from you! Feel free to reopen it or start a new one with updated details.

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and being part of the community!
